lib/tracing-limit/src/lib.rs

```diff
@@ -10,165 +10,200 @@ use std::{
 use tracing_core::{
     callsite::Identifier,
     field::{display, Field, Value, Visit},
+    span,
     subscriber::Interest,
     Event, Metadata, Subscriber,
 };
 use tracing_subscriber::layer::{Context, Layer};
 
-const RATE_LIMIT_FIELD: &str = "internal_log_rate_secs";
+const RATE_LIMIT_SECS_FIELD: &str = "internal_log_rate_secs";
+const RATE_LIMIT_KEY_FIELD: &str = "internal_log_rate_key";
+const COMPONENT_NAME_FIELD: &str = "component_name";
 const MESSAGE_FIELD: &str = "message";
-const DEFAULT_LIMIT: u64 = 5;
 
-#[derive(Debug, Default)]
-pub struct Limit {
-    events: RwLock<HashMap<Identifier, State>>,
-    callsite_store: RwLock<HashMap<Identifier, &'static Metadata<'static>>>,
+#[derive(Eq, PartialEq, Hash)]
+struct RateKeyIdentifier {
+    callsite: Identifier,
+    rate_limit_key: Option<String>,
+    component_name: Option<String>,
 }
 
-#[derive(Debug)]
-struct State {
-    start: Option<Instant>,
-    count: AtomicUsize,
-    limit: u64,
-    message: String,
-}
-
-impl Limit {
-    fn create_event<S: Subscriber>(
-        &self,
-        id: &Identifier,
-        ctx: &Context<S>,
-        message: String,
-        rate_limit: u64,
-    ) {
-        let store = self.callsite_store.read().unwrap();
-        let metadata = store.get(id).unwrap();
-
-        let fields = metadata.fields();
-
-        let message = display(message);
-
-        if let Some(message_field) = fields.field("message") {
-            let values = [(&message_field, Some(&message as &dyn Value))];
+pub struct RateLimitedLayer<S, L>
+where
+    L: Layer<S> + Sized,
+    S: Subscriber,
+{
+    events: RwLock<HashMap<RateKeyIdentifier, State>>,
+    inner: L,
 
-            let valueset = fields.value_set(&values);
-            let event = Event::new(metadata, &valueset);
-            ctx.event(&event);
-            drop(store);
-        } else {
-            let values = [(
-                &fields.field(RATE_LIMIT_FIELD).unwrap(),
-                Some(&rate_limit as &dyn Value),
-            )];
+    // TODO is this right?
+    _subscriber: std::marker::PhantomData<S>,
```

Comment:
I'm not familiar at all with this `PhantomData` marker. I simply followed rustc's recommendation to add it.

Reply:
All generic parameters have to be used in the struct. Since there is no actual data to correspond with `S`, you can add some `PhantomData` with that type. It's a zero-sized structure element that satisfies that use requirement.

Reply:
You'll also see these in structs where the data is there but Rust can't prove that it is. For instance, if you're doing `unsafe` style programming and you have a raw pointer to a something in memory that your struct "owns" but was not created by Rust (or was "forgotten" with with `mem::forget`). If your struct is parameterized on the type of that thing you need to use `PhantomData` to square the compiler's internal logic. 

Fun stuff.